core/lib/via_btc_client/src/signer/mod.rs

```diff
@@ -1 +1,360 @@
+
+        Ok(Self {
+            private_key,
+            rpc_client,
+        })
+    }
+
+    async fn sign_ecdsa(
+        &self,
+        unsigned_tx: &mut Transaction,
+        input_index: usize,
+    ) -> BitcoinSignerResult<()> {
+        let prevouts = self.get_prevouts(unsigned_tx).await?;
+        let mut sighash_cache = SighashCache::new(unsigned_tx.clone());
+        let secp = Secp256k1::new();
+
+        let secret_key = SecretKey::from_slice(&self.private_key.inner[..])
+            .map_err(|e| BitcoinError::SigningError(e.to_string()))?;
+
+        let sighash = sighash_cache
+            .p2wpkh_signature_hash(
+                input_index,
+                &prevouts[input_index].script_pubkey,
+                prevouts[input_index].value,
+                EcdsaSighashType::All,
+            )
+            .map_err(|e| BitcoinError::SigningError(e.to_string()))?;
+
+        let message = Message::from_digest_slice(sighash.as_ref())
+            .map_err(|e| BitcoinError::SigningError(e.to_string()))?;
+
+        let signature = secp.sign_ecdsa(&message, &secret_key);
+
+        let sig = bitcoin::ecdsa::Signature {
+            signature,
+            sighash_type: EcdsaSighashType::All,
+        };
+
+        let witness = Witness::p2wpkh(&sig, &self.private_key.inner.public_key(&secp));
+        unsigned_tx.input[input_index].witness = witness;
+
+        Ok(())
+    }
+
+    async fn sign_reveal(
+        &self,
+        unsigned_tx: &mut Transaction,
+        input_index: usize,
+        tapscript: &ScriptBuf,
+        leaf_version: LeafVersion,
+        control_block: &ControlBlock,
+    ) -> BitcoinSignerResult<()> {
+        let prevouts = self.get_prevouts(unsigned_tx).await?;
+        let mut sighash_cache = SighashCache::new(unsigned_tx.clone());
+        let secp = Secp256k1::new();
+
+        let secret_key = SecretKey::from_slice(&self.private_key.inner[..])
+            .map_err(|e| BitcoinError::SigningError(e.to_string()))?;
+        let keypair = Keypair::from_secret_key(&secp, &secret_key);
+
+        let reveal_sighash = sighash_cache
+            .taproot_script_spend_signature_hash(
+                input_index,
+                &Prevouts::All(&prevouts),
+                TapLeafHash::from_script(tapscript, leaf_version),
+                TapSighashType::All,
+            )
+            .map_err(|e| BitcoinError::SigningError(e.to_string()))?;
+
+        let reveal_msg = Message::from_digest(reveal_sighash.to_byte_array());
+        let reveal_signature = secp.sign_schnorr_no_aux_rand(&reveal_msg, &keypair);
+
+        let tap_sig = bitcoin::taproot::Signature {
+            signature: reveal_signature,
+            sighash_type: TapSighashType::All,
+        };
+
+        let mut reveal_witness = Witness::new();
+        reveal_witness.push(tap_sig.serialize());
+        reveal_witness.push(tapscript.as_bytes());
+        reveal_witness.push(control_block.serialize());
```

Comment:
the signer should only be responsible for signing, 
updating tx witness or working.

so it's better both method return the witness, instead of updating the tx via the mut reference